### PR TITLE
[ADD] base_currency_rate_exchange_method

### PR DESCRIPTION
**Pull request by andhit-r**
This module will give an option to use (1) direct method, or (2) indirect method for currency conversion. Odoo use indirect method for it currency conversion. Indirect method is hard to implement in country that has week currency exchange like Indonesia.

e.g:
- 1 USD = 13,833.00 IDR
- Using indirect method USD rate = 1 / 13.833,00 = 0.072290899
- Odoo currency rate only support 6 digit behind decimal. If we use indirect above we can only input it as 0.072290. This will cause inaccuracy.
- Using direct method, USD rate = 13,833.00

Indirect method is set to default (Odoo default conversion method)

### CODE REVIEW

**Comment by JordiBForgeFlow** [comment]
Hi @andhit-r I did not realize that you had proposed this module.

We have also implemented it, but using a different approach. See:
https://github.com/OCA/account-financial-tools/pull/337

The difference is that we maintain the currency conversion at the currency level. So you can define on a per-currency basis whether you want to apply the currency conversion or not. 

Still in countries like Indonesia, you might have a weak currency, but compared to some other currency like USD. But if you sell to Vietnamese Dong you would not want to apply currency conversion.

**Comment by andhit-r** [comment]
@jbeficent any solution for this kind of problem is ok for me. The reason why I maintain it at company level is purely for simpler usability. 

Let us ask community which more preferable.

**Comment by elicoidal** [comment]
@andhit-r 
My first impression was similar to yours: usually you apply one direction for all the currencies in one country (it is 1 EUR for any XX currency).
I dont see the business case for @jbeficent proposal but it includes your method and brings more flexibility for any company needing both approaches.

**Comment by elicoidal** [comment]
Anyway why not combining both modules into one with 2 possible setup at company level and currency level.
The result being the combination of both setup:
- direct in company + indirect in currency => indirect
- indirect in company + indirect in currency => direct
- etc.
  Flexible if necessary and simple for most of the cases (no need to maintain all currencies one by one).
  Side note comparing the modules, I think the one from @jbeficent is cleaner technically and I like the description of @andhit-r ones so working together we might achieve a great module :)

**Comment by andhit-r** [comment]
@elicoidal I agree with your idea to combine both approach. @jbeficent If you agree I don't mind combining both approach (using Your technical approach).

Case:
- Direct in company, inverse in currency == indirect
- Direct in company, no inverse in currency == direct
- Indirect in company, inverse in currency == direct
- Indirect in company, no inverse in currency == indirect

Company **currency exchange method** will be **required** and set to **indirect**. All currency inverse will be set to **False**

**Comment by andhit-r** [comment]
@pedrobaeza would you mind to put need review label for this PR.

**Comment by levkar** [comment]
I want to share our experience with currency rate modification that you might want to consider before making such a change.

1) hardwiring currency rate direct/indirect type to company (of the current user uid) is not a good idea. it might not work on some cases where such as workflows and cron jobs. how is it going to work in a multi company environment? 
2) it's best to pass these values by context but then you will need to fill your context whereever compute is called which means monkeypatching account module heavily. we made this for currency rate type, it's such a pain.
3) Make sure other addons are compatible with this (revaluation, currency rate update, etc.)
4) Currency is inverted also in our country but we prefer using system convention (1/inverted rate) but add (sometimes replace) inverted rate in another field. (ingadhoc's approach). 
5) When the currency rate is very low (or high in inverted), rounding/accuracy becomes a problem since you will need to divide 1 by the rate. You can overcome this by changing accuracy and rounding in currency rate.

In conclusion, I'm not sure if this is the way to go. Make sure you've considered points above and tested well before using this addon in production.

**Comment by andhit-r** [comment]
@levkar really appreciate Your input. Especially on point 1, i overlooked this important point. I will considered Your pointers and make adjustment. Closing this PR
